{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to read data from BigQuery\n",
    "\n",
    "This notebook demonstrates two ways to use BigQuery with Python\n",
    "1. by using SQL via [pandas-gbq](https://pandas-gbq.readthedocs.io/en/latest/)\n",
    "2. by using only Python code to extract the data of interest from BigQuery via [Ibis](https://docs.ibis-project.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting ibis-framework\n",
      "  Using cached ibis_framework-1.4.0-py3-none-any.whl (647 kB)\n",
      "Collecting pandas>=0.21\n",
      "  Using cached pandas-1.2.4-cp37-cp37m-manylinux1_x86_64.whl (9.9 MB)\n",
      "Collecting toolz\n",
      "  Using cached toolz-0.11.1-py3-none-any.whl (55 kB)\n",
      "Collecting pytz\n",
      "  Using cached pytz-2021.1-py2.py3-none-any.whl (510 kB)\n",
      "Collecting regex\n",
      "  Using cached regex-2021.4.4-cp37-cp37m-manylinux2014_x86_64.whl (720 kB)\n",
      "Collecting multipledispatch>=0.6.0\n",
      "  Using cached multipledispatch-0.6.0-py3-none-any.whl (11 kB)\n",
      "Collecting numpy>=1.11\n",
      "  Using cached numpy-1.20.3-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.3 MB)\n",
      "Collecting six\n",
      "  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)\n",
      "Collecting python-dateutil>=2.7.3\n",
      "  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)\n",
      "Installing collected packages: six, pytz, python-dateutil, numpy, toolz, regex, pandas, multipledispatch, ibis-framework\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "tensorflow-cpu 2.4.1 requires numpy~=1.19.2, but you have numpy 1.20.3 which is incompatible.\n",
      "tensorflow-cpu 2.4.1 requires six~=1.15.0, but you have six 1.16.0 which is incompatible.\n",
      "tensorflow-cpu 2.4.1 requires typing-extensions~=3.7.4, but you have typing-extensions 3.10.0.0 which is incompatible.\n",
      "cookiecutter 1.7.2 requires Jinja2<3.0.0, but you have jinja2 3.0.1 which is incompatible.\n",
      "cookiecutter 1.7.2 requires MarkupSafe<2.0.0, but you have markupsafe 2.0.1 which is incompatible.\u001b[0m\n",
      "Successfully installed ibis-framework-1.4.0 multipledispatch-0.6.0 numpy-1.20.3 pandas-1.2.4 python-dateutil-2.8.1 pytz-2021.1 regex-2021.4.4 six-1.16.0 toolz-0.11.1\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/dateutil already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/regex already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/pandas-1.2.4.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/python_dateutil-2.8.1.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/multipledispatch already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/six-1.16.0.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/numpy.libs already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/pandas already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/benchmarks already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/regex-2021.4.4.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/tlz already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/toolz-0.11.1.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/numpy already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/__pycache__ already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/six.py already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/numpy-1.20.3.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/pytz already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/ibis already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/multipledispatch-0.6.0.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/ibis_framework-1.4.0.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/toolz already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/pytz-2021.1.dist-info already exists. Specify --upgrade to force replacement.\u001b[0m\n",
      "\u001b[33mWARNING: Target directory /home/jupyter-user/notebooks/packages/bin already exists. Specify --upgrade to force replacement.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install ibis-framework"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pandas_gbq\n",
    "import ibis\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "BILLING_PROJECT_ID = os.environ['GOOGLE_PROJECT']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Option 1: Retrieve filtered data from BigQuery using SQL.\n",
    "\n",
    "The following SQL will read a subset of columns and subset of rows from a BigQuery table into a Pandas dataframe.\n",
    "* [Pandas](http://pandas.pydata.org/pandas-docs/stable/) is a popular Python package for data manipulation.\n",
    "* To learn more about SQL syntax see the [BigQuery standard SQL reference](https://cloud.google.com/bigquery/docs/reference/standard-sql/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info = pd.read_gbq(\"\"\"\n",
    "SELECT\n",
    "  Sample,\n",
    "  Gender,\n",
    "  Relationship,\n",
    "  Population,\n",
    "  Population_Description,\n",
    "  Super_Population,\n",
    "  Super_Population_Description,\n",
    "  Total_Exome_Sequence,\n",
    "  Main_Project_E_Platform,\n",
    "  Main_Project_E_Centers\n",
    "FROM\n",
    "  `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\n",
    "WHERE\n",
    "  -- Only include information for samples in phase 1.\n",
    "  In_Phase1_Integrated_Variant_Set = TRUE\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1092 entries, 0 to 1091\n",
      "Data columns (total 10 columns):\n",
      " #   Column                        Non-Null Count  Dtype  \n",
      "---  ------                        --------------  -----  \n",
      " 0   Sample                        1092 non-null   object \n",
      " 1   Gender                        1092 non-null   object \n",
      " 2   Relationship                  1092 non-null   object \n",
      " 3   Population                    1092 non-null   object \n",
      " 4   Population_Description        1092 non-null   object \n",
      " 5   Super_Population              1092 non-null   object \n",
      " 6   Super_Population_Description  1092 non-null   object \n",
      " 7   Total_Exome_Sequence          1069 non-null   float64\n",
      " 8   Main_Project_E_Platform       1092 non-null   object \n",
      " 9   Main_Project_E_Centers        1092 non-null   object \n",
      "dtypes: float64(1), object(9)\n",
      "memory usage: 85.4+ KB\n"
     ]
    }
   ],
   "source": [
    "sample_info.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total_Exome_Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.069000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.028243e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>3.845551e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>2.477265e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>8.078811e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>9.730942e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.142406e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3.511709e+10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Total_Exome_Sequence\n",
       "count          1.069000e+03\n",
       "mean           1.028243e+10\n",
       "std            3.845551e+09\n",
       "min            2.477265e+09\n",
       "25%            8.078811e+09\n",
       "50%            9.730942e+09\n",
       "75%            1.142406e+10\n",
       "max            3.511709e+10"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>Population</th>\n",
       "      <th>Population_Description</th>\n",
       "      <th>Super_Population</th>\n",
       "      <th>Super_Population_Description</th>\n",
       "      <th>Total_Exome_Sequence</th>\n",
       "      <th>Main_Project_E_Platform</th>\n",
       "      <th>Main_Project_E_Centers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00377</td>\n",
       "      <td>female</td>\n",
       "      <td></td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>4.105720e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG01278</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>CLM</td>\n",
       "      <td>Colombian in Medellin, Colombia</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>7.717205e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BCM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA18527</td>\n",
       "      <td>female</td>\n",
       "      <td></td>\n",
       "      <td>CHB</td>\n",
       "      <td>Han Chinese in Bejing, China</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>5.811252e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BCM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA19672</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA20363</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>ASW</td>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample  Gender Relationship Population  \\\n",
       "0  HG00377  female                     FIN   \n",
       "1  HG01278  female       mother        CLM   \n",
       "2  NA18527  female                     CHB   \n",
       "3  NA19672  female       mother        MXL   \n",
       "4  NA20363  female       mother        ASW   \n",
       "\n",
       "                        Population_Description Super_Population  \\\n",
       "0                           Finnish in Finland              EUR   \n",
       "1              Colombian in Medellin, Colombia              AMR   \n",
       "2                 Han Chinese in Bejing, China              EAS   \n",
       "3  Mexican Ancestry in Los Angeles, California              AMR   \n",
       "4             African Ancestry in Southwest US              AFR   \n",
       "\n",
       "  Super_Population_Description  Total_Exome_Sequence Main_Project_E_Platform  \\\n",
       "0                     European          4.105720e+09                ILLUMINA   \n",
       "1                     American          7.717205e+09                ILLUMINA   \n",
       "2                   East Asian          5.811252e+09                ILLUMINA   \n",
       "3                     American                   NaN                           \n",
       "4                      African                   NaN                           \n",
       "\n",
       "  Main_Project_E_Centers  \n",
       "0                     BI  \n",
       "1                    BCM  \n",
       "2                    BCM  \n",
       "3                         \n",
       "4                         "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Option 2: Retrieve filtered data from BigQuery using Python.\n",
    "\n",
    "The following Python code will read a BigQuery table into a Pandas dataframe.\n",
    "\n",
    "From https://cloud.google.com/community/tutorials/bigquery-ibis\n",
    "\n",
    "*[Ibis](http://ibis-project.org/) is a Python library for doing data analysis. It offers a Pandas-like environment for executing data analysis in big data processing systems such as Google BigQuery. Ibis's primary goals are to be a type safe, expressive, composable, and familiar replacement for SQL.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = ibis.bigquery.connect(\n",
    "    project_id=BILLING_PROJECT_ID,\n",
    "    dataset_id='bigquery-public-data.human_genome_variants')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ibis version: 1.4.0\n"
     ]
    }
   ],
   "source": [
    "print('ibis version: %s' % ibis.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BigQueryTable[table]\n",
       "  name: bigquery-public-data.human_genome_variants.1000_genomes_sample_info\n",
       "  schema:\n",
       "    Sample : string\n",
       "    Family_ID : string\n",
       "    Population : string\n",
       "    Population_Description : string\n",
       "    Gender : string\n",
       "    Relationship : string\n",
       "    Unexpected_Parent_Child : string\n",
       "    Non_Paternity : string\n",
       "    Siblings : string\n",
       "    Grandparents : string\n",
       "    Avuncular : string\n",
       "    Half_Siblings : string\n",
       "    Unknown_Second_Order : string\n",
       "    Third_Order : string\n",
       "    In_Low_Coverage_Pilot : boolean\n",
       "    LC_Pilot_Platforms : string\n",
       "    LC_Pilot_Centers : string\n",
       "    In_High_Coverage_Pilot : boolean\n",
       "    HC_Pilot_Platforms : string\n",
       "    HC_Pilot_Centers : string\n",
       "    In_Exon_Targetted_Pilot : boolean\n",
       "    ET_Pilot_Platforms : string\n",
       "    ET_Pilot_Centers : string\n",
       "    Has_Sequence_in_Phase1 : boolean\n",
       "    Phase1_LC_Platform : string\n",
       "    Phase1_LC_Centers : string\n",
       "    Phase1_E_Platform : string\n",
       "    Phase1_E_Centers : string\n",
       "    In_Phase1_Integrated_Variant_Set : boolean\n",
       "    Has_Phase1_chrY_SNPS : boolean\n",
       "    Has_phase1_chrY_Deletions : boolean\n",
       "    Has_phase1_chrMT_SNPs : boolean\n",
       "    Main_project_LC_Centers : string\n",
       "    Main_project_LC_platform : string\n",
       "    Total_LC_Sequence : float64\n",
       "    LC_Non_Duplicated_Aligned_Coverage : float64\n",
       "    Main_Project_E_Centers : string\n",
       "    Main_Project_E_Platform : string\n",
       "    Total_Exome_Sequence : float64\n",
       "    X_Targets_Covered_to_20x_or_greater : float64\n",
       "    VerifyBam_E_Omni_Free : float64\n",
       "    VerifyBam_E_Affy_Free : float64\n",
       "    VerifyBam_E_Omni_Chip : float64\n",
       "    VerifyBam_E_Affy_Chip : float64\n",
       "    VerifyBam_LC_Omni_Free : float64\n",
       "    VerifyBam_LC_Affy_Free : float64\n",
       "    VerifyBam_LC_Omni_Chip : float64\n",
       "    VerifyBam_LC_Affy_Chip : float64\n",
       "    LC_Indel_Ratio : float64\n",
       "    E_Indel_Ratio : float64\n",
       "    LC_Passed_QC : boolean\n",
       "    E_Passed_QC : boolean\n",
       "    In_Final_Phase_Variant_Calling : boolean\n",
       "    Has_Omni_Genotypes : boolean\n",
       "    Has_Axiom_Genotypes : boolean\n",
       "    Has_Affy_6_0_Genotypes : boolean\n",
       "    Has_Exome_LOF_Genotypes : boolean\n",
       "    EBV_Coverage : float64\n",
       "    DNA_Source_from_Coriell : string\n",
       "    Has_Sequence_from_Blood_in_Index : boolean\n",
       "    Super_Population : string\n",
       "    Super_Population_Description : string"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_tbl = conn.table('1000_genomes_sample_info')\n",
    "sample_info_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the filter criteria.\n",
    "phase_1_only = sample_info_tbl.In_Phase1_Integrated_Variant_Set == True\n",
    "\n",
    "# Apply the filter and choose the columns to return.\n",
    "phase_1_sample_info_tbl = sample_info_tbl.filter(phase_1_only)['Sample', 'Gender', 'Relationship', 'Population', 'Population_Description',\n",
    "                'Super_Population', 'Super_Population_Description', 'Total_Exome_Sequence',\n",
    "                'Main_Project_E_Platform', 'Main_Project_E_Centers']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT `Sample`, `Gender`, `Relationship`, `Population`,\n",
      "       `Population_Description`, `Super_Population`,\n",
      "       `Super_Population_Description`, `Total_Exome_Sequence`,\n",
      "       `Main_Project_E_Platform`, `Main_Project_E_Centers`\n",
      "FROM (\n",
      "  SELECT *\n",
      "  FROM `bigquery-public-data.human_genome_variants.1000_genomes_sample_info`\n",
      "  WHERE `In_Phase1_Integrated_Variant_Set` = TRUE\n",
      ") t0\n"
     ]
    }
   ],
   "source": [
    "# Optional: take a look at the SQL.\n",
    "print(phase_1_sample_info_tbl.compile())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/dist-packages/google/cloud/bigquery/client.py:462: UserWarning: Cannot create BigQuery Storage client, the dependency google-cloud-bigquery-storage is not installed.\n",
      "  \"Cannot create BigQuery Storage client, the dependency \"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1092"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Optional: See how much data this will return.\n",
    "phase_1_sample_info_tbl.count().execute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1092, 10)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Go ahead and retrieve the data.\n",
    "phase_1_sample_info_df = phase_1_sample_info_tbl.limit(1000000).execute()\n",
    "phase_1_sample_info_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>Population</th>\n",
       "      <th>Population_Description</th>\n",
       "      <th>Super_Population</th>\n",
       "      <th>Super_Population_Description</th>\n",
       "      <th>Total_Exome_Sequence</th>\n",
       "      <th>Main_Project_E_Platform</th>\n",
       "      <th>Main_Project_E_Centers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00377</td>\n",
       "      <td>female</td>\n",
       "      <td></td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>4.105720e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG01278</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>CLM</td>\n",
       "      <td>Colombian in Medellin, Colombia</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>7.717205e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BCM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA18527</td>\n",
       "      <td>female</td>\n",
       "      <td></td>\n",
       "      <td>CHB</td>\n",
       "      <td>Han Chinese in Bejing, China</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>5.811252e+09</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>BCM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA19672</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA20363</td>\n",
       "      <td>female</td>\n",
       "      <td>mother</td>\n",
       "      <td>ASW</td>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample  Gender Relationship Population  \\\n",
       "0  HG00377  female                     FIN   \n",
       "1  HG01278  female       mother        CLM   \n",
       "2  NA18527  female                     CHB   \n",
       "3  NA19672  female       mother        MXL   \n",
       "4  NA20363  female       mother        ASW   \n",
       "\n",
       "                        Population_Description Super_Population  \\\n",
       "0                           Finnish in Finland              EUR   \n",
       "1              Colombian in Medellin, Colombia              AMR   \n",
       "2                 Han Chinese in Bejing, China              EAS   \n",
       "3  Mexican Ancestry in Los Angeles, California              AMR   \n",
       "4             African Ancestry in Southwest US              AFR   \n",
       "\n",
       "  Super_Population_Description  Total_Exome_Sequence Main_Project_E_Platform  \\\n",
       "0                     European          4.105720e+09                ILLUMINA   \n",
       "1                     American          7.717205e+09                ILLUMINA   \n",
       "2                   East Asian          5.811252e+09                ILLUMINA   \n",
       "3                     American                   NaN                           \n",
       "4                      African                   NaN                           \n",
       "\n",
       "  Main_Project_E_Centers  \n",
       "0                     BI  \n",
       "1                    BCM  \n",
       "2                    BCM  \n",
       "3                         \n",
       "4                         "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phase_1_sample_info_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Provenance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-19 23:15:19.915762\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "absl-py==0.12.0\r\n",
      "anyio==3.1.0\r\n",
      "argon2-cffi==20.1.0\r\n",
      "arrow==1.1.0\r\n",
      "arviz==0.11.2\r\n",
      "asn1crypto==0.24.0\r\n",
      "astroid==2.5.6\r\n",
      "astunparse==1.6.3\r\n",
      "async-generator==1.10\r\n",
      "attrs==21.2.0\r\n",
      "backcall==0.2.0\r\n",
      "bagit==1.8.1\r\n",
      "bgzip==0.3.5\r\n",
      "binaryornot==0.4.4\r\n",
      "biopython==1.78\r\n",
      "bleach==3.3.0\r\n",
      "bokeh==2.3.1\r\n",
      "brewer2mpl==1.4.1\r\n",
      "bx-python==0.8.11\r\n",
      "CacheControl==0.11.7\r\n",
      "cachetools==4.2.2\r\n",
      "certifi==2020.12.5\r\n",
      "cffi==1.14.5\r\n",
      "cftime==1.4.1\r\n",
      "chardet==4.0.0\r\n",
      "cli-builder==0.1.5\r\n",
      "click==7.1.2\r\n",
      "colorama==0.4.4\r\n",
      "confuse==1.4.0\r\n",
      "cookiecutter==1.7.2\r\n",
      "crcmod==1.7\r\n",
      "cryptography==3.4.7\r\n",
      "cwltool==1.0.20190228155703\r\n",
      "cycler==0.10.0\r\n",
      "Cython==0.29.23\r\n",
      "decorator==4.4.2\r\n",
      "defusedxml==0.7.1\r\n",
      "descartes==1.1.0\r\n",
      "dill==0.3.3\r\n",
      "entrypoints==0.3\r\n",
      "facets-overview==1.0.0\r\n",
      "fastinterval==0.1.1\r\n",
      "fastprogress==1.0.0\r\n",
      "filelock==3.0.12\r\n",
      "firecloud==0.16.25\r\n",
      "flatbuffers==1.12\r\n",
      "future==0.18.2\r\n",
      "gast==0.3.3\r\n",
      "ggplot==0.11.5\r\n",
      "gitdb==4.0.7\r\n",
      "GitPython==3.1.17\r\n",
      "google-api-core==1.26.3\r\n",
      "google-auth==1.30.0\r\n",
      "google-auth-oauthlib==0.4.4\r\n",
      "google-cloud-bigquery==2.15.0\r\n",
      "google-cloud-bigquery-datatransfer==3.1.1\r\n",
      "google-cloud-core==1.6.0\r\n",
      "google-cloud-datastore==2.1.2\r\n",
      "google-cloud-resource-manager==0.30.3\r\n",
      "google-cloud-storage==1.38.0\r\n",
      "google-crc32c==1.1.2\r\n",
      "google-pasta==0.2.0\r\n",
      "google-resumable-media==1.2.0\r\n",
      "googleapis-common-protos==1.53.0\r\n",
      "grpcio==1.32.0\r\n",
      "gs-chunked-io==0.5.2\r\n",
      "h5py==2.10.0\r\n",
      "html5lib==1.1\r\n",
      "htmlmin==0.1.12\r\n",
      "httplib2==0.19.1\r\n",
      "ibis-framework==1.4.0\r\n",
      "idna==2.10\r\n",
      "ImageHash==4.2.0\r\n",
      "importlib-metadata==4.0.1\r\n",
      "intel-openmp==2021.2.0\r\n",
      "ipykernel==5.5.4\r\n",
      "ipython==7.23.1\r\n",
      "ipython-genutils==0.2.0\r\n",
      "ipywidgets==7.6.3\r\n",
      "isodate==0.6.0\r\n",
      "isort==5.8.0\r\n",
      "jedi==0.18.0\r\n",
      "Jinja2==3.0.1\r\n",
      "jinja2-time==0.2.0\r\n",
      "joblib==1.0.1\r\n",
      "jsonschema==3.2.0\r\n",
      "jupyter==1.0.0\r\n",
      "jupyter-client==6.1.12\r\n",
      "jupyter-console==6.4.0\r\n",
      "jupyter-contrib-core==0.3.3\r\n",
      "jupyter-contrib-nbextensions==0.5.1\r\n",
      "jupyter-core==4.7.1\r\n",
      "jupyter-highlight-selected-word==0.2.0\r\n",
      "jupyter-latex-envs==1.4.6\r\n",
      "jupyter-nbextensions-configurator==0.4.1\r\n",
      "jupyter-server==1.7.0\r\n",
      "jupyter-server-mathjax==0.2.2\r\n",
      "jupyterlab==0.35.4\r\n",
      "jupyterlab-pygments==0.1.2\r\n",
      "jupyterlab-server==0.2.0\r\n",
      "jupyterlab-widgets==1.0.0\r\n",
      "Keras==2.4.3\r\n",
      "Keras-Preprocessing==1.1.2\r\n",
      "keyring==10.6.0\r\n",
      "keyrings.alt==3.0\r\n",
      "kiwisolver==1.3.1\r\n",
      "lazy-object-proxy==1.6.0\r\n",
      "libcst==0.3.18\r\n",
      "lockfile==0.12.2\r\n",
      "lxml==4.6.3\r\n",
      "Mako==1.1.4\r\n",
      "Markdown==3.3.4\r\n",
      "MarkupSafe==2.0.1\r\n",
      "matplotlib==3.4.1\r\n",
      "matplotlib-inline==0.1.2\r\n",
      "matplotlib-venn==0.11.6\r\n",
      "mccabe==0.6.1\r\n",
      "missingno==0.4.2\r\n",
      "mistune==0.8.4\r\n",
      "mizani==0.7.3\r\n",
      "mkl==2021.2.0\r\n",
      "multipledispatch==0.6.0\r\n",
      "mypy-extensions==0.4.3\r\n",
      "nbclient==0.5.3\r\n",
      "nbconvert==6.0.7\r\n",
      "nbdime==3.0.0\r\n",
      "nbformat==5.1.3\r\n",
      "nest-asyncio==1.5.1\r\n",
      "netCDF4==1.5.6\r\n",
      "networkx==2.5.1\r\n",
      "nose==1.3.7\r\n",
      "notebook==6.1.1\r\n",
      "numpy==1.20.3\r\n",
      "oauth2client==4.1.3\r\n",
      "oauthlib==3.1.0\r\n",
      "opt-einsum==3.3.0\r\n",
      "packaging==20.9\r\n",
      "palettable==3.3.0\r\n",
      "pandas==1.2.4\r\n",
      "pandas-gbq==0.15.0\r\n",
      "pandas-profiling==2.12.0\r\n",
      "pandocfilters==1.4.3\r\n",
      "parso==0.8.2\r\n",
      "patsy==0.5.1\r\n",
      "pdoc3==0.9.2\r\n",
      "pexpect==4.8.0\r\n",
      "phik==0.11.2\r\n",
      "pickleshare==0.7.5\r\n",
      "Pillow==8.2.0\r\n",
      "plotnine==0.8.0\r\n",
      "poyo==0.5.0\r\n",
      "prometheus-client==0.10.1\r\n",
      "prompt-toolkit==3.0.18\r\n",
      "proto-plus==1.18.1\r\n",
      "protobuf==3.17.0\r\n",
      "prov==1.5.1\r\n",
      "psutil==5.8.0\r\n",
      "ptyprocess==0.7.0\r\n",
      "py4j==0.10.9.2\r\n",
      "pyarrow==3.0.0\r\n",
      "pyasn1==0.4.8\r\n",
      "pyasn1-modules==0.2.8\r\n",
      "pycparser==2.20\r\n",
      "pycrypto==2.6.1\r\n",
      "pydata-google-auth==1.2.0\r\n",
      "pydot==1.4.2\r\n",
      "pyfasta==0.5.2\r\n",
      "Pygments==2.9.0\r\n",
      "pygobject==3.26.1\r\n",
      "pylint==1.7.2\r\n",
      "pymc3==3.11.2\r\n",
      "pyOpenSSL==20.0.1\r\n",
      "pyparsing==2.4.7\r\n",
      "pyrsistent==0.17.3\r\n",
      "pysam==0.16.0.1\r\n",
      "python-apt==1.6.5+ubuntu0.5\r\n",
      "python-datauri==1.0.0\r\n",
      "python-dateutil==2.8.1\r\n",
      "python-lzo==1.12\r\n",
      "python-slugify==5.0.2\r\n",
      "pytz==2021.1\r\n",
      "PyVCF==0.6.8\r\n",
      "PyWavelets==1.1.1\r\n",
      "pyxdg==0.25\r\n",
      "PyYAML==5.4.1\r\n",
      "pyzmq==22.0.3\r\n",
      "qtconsole==5.1.0\r\n",
      "QtPy==1.9.0\r\n",
      "rdflib==4.2.2\r\n",
      "rdflib-jsonld==0.4.0\r\n",
      "readline==6.2.4.1\r\n",
      "regex==2021.4.4\r\n",
      "requests==2.25.1\r\n",
      "requests-oauthlib==1.3.0\r\n",
      "rsa==4.7.2\r\n",
      "ruamel.yaml==0.15.77\r\n",
      "scandir==1.10.0\r\n",
      "schema-salad==4.5.20191229160203\r\n",
      "scikit-learn==0.24.2\r\n",
      "scipy==1.6.3\r\n",
      "seaborn==0.11.1\r\n",
      "SecretStorage==2.3.1\r\n",
      "semver==2.13.0\r\n",
      "Send2Trash==1.5.0\r\n",
      "shellescape==3.4.1\r\n",
      "six==1.16.0\r\n",
      "smmap==4.0.0\r\n",
      "sniffio==1.2.0\r\n",
      "statsmodels==0.12.2\r\n",
      "tangled-up-in-unicode==0.0.7\r\n",
      "tbb==2021.2.0\r\n",
      "tensorboard==2.5.0\r\n",
      "tensorboard-data-server==0.6.1\r\n",
      "tensorboard-plugin-wit==1.8.0\r\n",
      "tensorflow-cpu==2.4.1\r\n",
      "tensorflow-estimator==2.4.0\r\n",
      "termcolor==1.1.0\r\n",
      "terminado==0.10.0\r\n",
      "terra-notebook-utils==0.7.0\r\n",
      "testpath==0.5.0\r\n",
      "text-unidecode==1.3\r\n",
      "Theano==1.0.5\r\n",
      "Theano-PyMC==1.1.2\r\n",
      "threadpoolctl==2.1.0\r\n",
      "toolz==0.11.1\r\n",
      "tornado==6.1\r\n",
      "tqdm==4.60.0\r\n",
      "traitlets==5.0.5\r\n",
      "typed-ast==1.4.3\r\n",
      "typing-extensions==3.10.0.0\r\n",
      "typing-inspect==0.6.0\r\n",
      "unattended-upgrades==0.1\r\n",
      "urllib3==1.26.4\r\n",
      "visions==0.6.0\r\n",
      "wcwidth==0.2.5\r\n",
      "webencodings==0.5.1\r\n",
      "websocket-client==1.0.0\r\n",
      "Werkzeug==1.0.1\r\n",
      "widgetsnbextension==3.5.1\r\n",
      "wrapt==1.12.1\r\n",
      "xarray==0.17.0\r\n",
      "zipp==3.4.1\r\n"
     ]
    }
   ],
   "source": [
    "!pip3 freeze"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copyright 2018 The Broad Institute, Inc., Verily Life Sciences, LLC All rights reserved.\n",
    "\n",
    "This software may be modified and distributed under the terms of the BSD license. See the LICENSE file for details."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "336px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
